sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -925,18 +925,20 @@ public boolean start() throws IOException {
       consumer = source.consumerFactoryFn.apply(source.consumerConfig);
       consumer.assign(source.assignedPartitions);
 
-      // seek to consumedOffset + 1 if it is set
       for (PartitionState p : partitionStates) {
-        if (p.consumedOffset >= 0) {
-          LOG.info("{}: resuming {} at {}", name, p.topicPartition, p.consumedOffset + 1);
+        if (p.consumedOffset >=  0) {
           consumer.seek(p.topicPartition, p.consumedOffset + 1);
         } else {
-          LOG.info("{}: resuming {} at default offset", name, p.topicPartition);
+          // set consumed offset to (next offset - 1), otherwise checkpoint would contain invalid
+          // offset until we read first record from this partition.
+          p.consumedOffset = consumer.position(p.topicPartition) - 1;
```

Comment:
Good point. I updated the patch. Now we have two negative offset ('UNINITIALIZED' and 'EMPTY_PARTITION').

We wouldn't need two of these if we tracked next offset rather than consumedOffsets. May be we should do that.